### PR TITLE
Allow component modules to be loaded from any path

### LINKED ISSUE #116: hasFile() prevents component modules from being relocated
The expected location for externally packaged Perl modules varies with distribution and may change in future releases:
* EL6 and EL7 — [`/usr/share/perl5/vendor_perl`](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7/html/Developer_Guide/lib.details.html#idp14646592)
* Debian (and derivatives) — [`/usr/share/perl5`](https://www.debian.org/doc/packaging-manuals/perl-policy/ch-perl.html#s-paths)

Our current implementation [assumes](https://github.com/quattor/ncm-ncd/blob/0028c74638b76083f54e1b8b2bf1a03752281306/src/main/perl/ComponentProxy.pm#L26) that component modules are located in `/usr/lib/perl/NCM/Component`, while this is the case if the components are installed by our RPMs, it may not be if they are installed by other means.

Removing the call to `hasFile` in `_load` allows the module to be loaded from any of the standard paths without issue. I assume the original intent here was to detect a missing module before is was loaded, but the `eval` block that loads the module deals with that case.

For example, with the check removed and the dirperm component missing:
```
# quattor-configure --nodeps --noautodeps dirperm accounts

[INFO] NCM-NCD version 17.3.0 started by root at: Tue Jun 27 13:25:38 2017
[INFO] Using profile CID 1 name prod_cloud-domain
[INFO] Ignoring broken pre/post dependencies (--nodeps flag set)
[INFO] executing configure on components....


[INFO] running component: dirperm
---------------------------------------------------------
[ERROR] bad Perl code in NCM::Component::dirperm (/usr/lib/perl/NCM/Component/dirperm.pm): Can't locate NCM/Component/dirperm.pm in @INC (you may need to install the NCM::Component::dirperm module) (@INC contains: /usr/lib/perl /etc/perl /usr/local/lib/perl/5.18.2 /usr/local/share/perl/5.18.2 /usr/lib/perl5 /usr/share/perl5 /usr/lib/perl/5.18 /usr/share/perl/5.18 /usr/local/lib/site_perl) at /usr/share/perl/5.18/Module/Load.pm line 27.
Can't locate NCM/Component/dirperm in @INC (@INC contains: /usr/lib/perl /etc/perl /usr/local/lib/perl/5.18.2 /usr/local/share/perl/5.18.2 /usr/lib/perl5 /usr/share/perl5 /usr/lib/perl/5.18 /usr/share/perl/5.18 /usr/local/lib/site_perl) at /usr/share/perl/5.18/Module/Load.pm line 27.

[ERROR] cannot load component: dirperm
[ERROR] cannot execute configure on component dirperm

[INFO] running component: accounts
---------------------------------------------------------
[INFO] System has 28 accounts in 52 groups
[INFO] Committing 52 groups
[INFO] Committing 28 accounts
[INFO] configure on component accounts executed, 0 errors, 0 warnings

=========================================================

[INFO] Errors while configuringing dirperm (1)
[ERROR] 1 errors, 0 warnings executing configure
```

I'll open a PR shortly which removes `hasFile()` so we can discuss.

### PR DESCRIPTION
**Pull request by jrha**
Remove hasFile method which artificially restricts the paths which component
modules can be loaded from, allowing loading from any valid path in `@INC`.

This allows component modules to be placed in a location appropriate for each
operating system environments we target.

This change breaks backwards compatability by:
* Removing the **file?** column from the output of `reportComponents`
  * Used by `ncm-ncd --list`/`quattor-list` so could *potentially* be being parsed by someone.
* Removing `hasFile` entirely.
* Removing `getComponentFilename` entirely.
* Allowing modules to be loaded from other paths.
  * Potentially resulting in different/unexpected versions of code being run.

Resolves #116.

### CODE REVIEW

**Comment by stdweird** [comment]
@jrha i don't know the original reason not to use the perl system path, but this is against that (obviously).

it also allows to pick up any component from PERL5LIB env variable, not sure that is an issue.

we can probably add a restriction to also consider the perl builtin INC paths, and not only `/usr/lib/perl`

also, please add unittests

**Comment by jrha** [comment]
I guess @piojo might have some insight into the benefits of path restrictions, I assumed that these were just kludges left over from pre EL6 days. I'll sort out the tests later, just wanted to get this out so we could discuss it.

File: src/main/perl/ComponentProxy.pm [hunk 1]
```diff
@@ -303,9 +278,10 @@ sub _load
     eval {
         local $SIG{__WARN__} = sub { push(@warns, $_[0]); };
         load $package;
+        $mod_fn = $self->_getComponentFilename();
```
Comment (stdweird) [inline src/main/perl/ComponentProxy.pm]: remove this from the eval, run this after the next if block, incl the `my $mod_fn = '';` from above

File: src/main/perl/ComponentProxy.pm [hunk 2]
```diff
@@ -288,13 +267,9 @@ sub _load
...
     my $name = $self->name();
-
-    my $mod_fn = $self->getComponentFilename();
-    if (!$self->hasFile()) {
-        # No arguments passed to hasFile
-        $self->error("component $mod is not installed (looking for $mod_fn)");
-        return;
-    }
+    my $mod_fn = '';
+    $self->debug(5, "Component module is $mod");
```
Comment (stdweird) [inline src/main/perl/ComponentProxy.pm]: single debug statement with both module and name

**Comment by jrha** [comment]
Make sure I squash these commits before this gets merged.

**Comment by stdweird** [comment]
we agreed during the workshop of april 2018 to allow this, and this will be merged once reviewed

File: src/main/perl/ComponentProxy.pm [hunk 3]
```diff
@@ -326,7 +272,7 @@ sub _load
         $self->error('bad component exception handler: $EC is not defined, ',
                      'not accessible or not of type LC::Exception::Context',
                      "(note 1: the component package name has to be exactly ",
-                     "'$package' - please verify this inside $mod_fn) ",
+                     "'$package' - please verify this in the source code) ",
```
Comment (stdweird) [inline src/main/perl/ComponentProxy.pm]: add `of $package ($mod)` to make it clearer what was loaded/used

**Comment by stdweird** [comment]
@jrha just a minor remark, this is otherwise good to go

**Comment by jrha** [comment]
Well I broke it, will try and figure out what I did later.

**Comment by jrha** [comment]
@hpcugentbot please retest this please

**Comment by jrha** [comment]
Removed the test checking internal function before component initialised that was causing the failure.

**Comment by jrha** [comment]
@stdweird could you re-review this when you have a moment? :innocent:

**Comment by stdweird** [comment]
@ned21 @jouvin i'll let you merge this so you aware of this new behaviour
@jrha thx for fixing this!

**Comment by ned21** [comment]
I suspect this was originally in place because of CVE-2016-1238 https://rt.perl.org/Public/Bug/Display.html?id=127834 has more details but it's related to . in `@INC`.  So I'm not sure we can or should trust `@INC` even in [EL7](https://bugzilla.redhat.com/show_bug.cgi?id=1355695).

Can we block any modules being loaded from `.` before merging this?

**Comment by stdweird** [comment]
@ned21 you want that filtered here? i tought we already did this in the client tools (because otherwise you could load `NCM::Component` etc from `.`)?

**Comment by ned21** [comment]
I thought I remembered that conversation too. So maybe that mitigation means it's now safe to remove from here? Can we confirm where the protection is now implemented and get that documented either inline as a comment or in the commit log?

**Comment by stdweird** [comment]
@ned21 for ncm-ncd this was done in https://github.com/quattor/ncm-ncd/pull/123/commits/5cfe99375bae66cd176fdad8629ae44534d4a51e

**Comment by jrha** [comment]
@stdweird @ned21 commit message updated with explanation about CVE-2016-1238

**Pull request merged**